traversal/common/Modifiers.java

```diff
@@ -0,0 +1,144 @@
+        return Objects.equals(filter, other.filter) && Objects.equals(sorting, other.sorting);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(filter, sorting);
+    }
+
+    public static class Filter {
+
+        Set<Identifier.Variable.Retrievable> variables;
+
+        private Filter(Set<Identifier.Variable.Retrievable> variables) {
+            this.variables = variables;
+        }
+
+        public static Filter create(Set<Identifier.Variable.Retrievable> variables) {
+            return new Filter(variables);
+        }
+
+        public static Filter create(List<UnboundVariable> vars) {
+            Set<Identifier.Variable.Retrievable> variables = new HashSet<>();
+            iterate(vars).map(v -> Identifier.Variable.of(v.reference().asName())).forEachRemaining(variables::add);
+            return new Filter(variables);
+        }
+
+        public Set<Identifier.Variable.Retrievable> variables() {
+            return variables;
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+            Filter other = (Filter) o;
+            return variables.equals(other.variables);
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(variables);
+        }
+    }
+
+    public static class Sorting {
+
+        public static Sorting EMPTY = new Sorting(list(), map());
+
+        List<Identifier.Variable.Retrievable> variables;
+        Map<Identifier.Variable.Retrievable, Boolean> ascending;
+
+        private Sorting(List<Identifier.Variable.Retrievable> variables, Map<Identifier.Variable.Retrievable, Boolean> ascending) {
+            assert variables.size() == ascending.size() && variables.containsAll(ascending.keySet());
+            this.variables = variables;
+            this.ascending = ascending;
+        }
+
+        public static Sorting create(Sortable.Sorting sort) {
+            List<Identifier.Variable.Retrievable> variables = new ArrayList<>();
+            Map<Identifier.Variable.Retrievable, Boolean> ascending = new HashMap<>();
+            // TODO: sort order per-variable
+            sort.vars().forEach(typeQLVar -> {
+                Identifier.Variable.Retrievable var = Identifier.Variable.of(typeQLVar.reference().asReferable()).asRetrievable();
+                variables.add(var);
+                ascending.put(var, sort.order() == TypeQLArg.Order.ASC);
+            });
+            return new Sorting(variables, ascending);
+        }
+
+        public List<Identifier.Variable.Retrievable> variables() {
+            return variables;
+        }
+
+        public boolean isAscending(Identifier id) {
+            if (!id.isRetrievable() || !ascending.containsKey(id.asVariable().asRetrievable())) return true;
+            return ascending.get(id.asVariable().asRetrievable());
```

Comment:
note: we're encoding a key assumtion here that anything that is not listed or not explicitly descending is ascending!